hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java

```diff
@@ -690,6 +690,14 @@ public void foundCircularAssociation(AssociationAttributeDefinition attributeDef
 		}
 	}
 
+	private Joinable locateJoinable(AttributeSource source) {
+		if ( source instanceof CompositionDefinition ) {
+			return locateJoinable( ( (CompositionDefinition) source ).getSource() );
+		}
+
+		return (Joinable) ( (EntityDefinition) source ).getEntityPersister();
```

Comment:
I don't know much about this code section, but it seems a bit suspicious that the "else" branch here is taking an entirely different approach than the current code?

It seems that the bugfix should only affect the case in which is it a `CompositionDefinition`, and it makes me a bit uncomfortable to change the other branch as well without understanding more about the reasons.

Reply:
@Sanne I have worked on a different solution https://github.com/hibernate/hibernate-orm/pull/3635